File: backend/app/Models/RawMaterial.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class RawMaterial extends Model
{
    protected $fillable = ['name', 'stock', 'unit', 'min_stock', 'is_active', 'image'];

    protected $casts = [
        'stock' => 'decimal:2',
        'min_stock' => 'decimal:2',
        'is_active' => 'boolean',
    ];

    /**
     * Adjust stock for a raw material and record the log.
     */
    public static function adjustStock(
        int $rawMaterialId,
        float $quantity,
        string $type = 'adjustment',
        ?string $reason = null,
        ?string $notes = null,
        ?int $userId = null
    ): static {
        $material = static::findOrFail($rawMaterialId);
        $stockBefore = (float) $material->stock;
        $stockAfter = max(0, $stockBefore + $quantity);

        $material->update(['stock' => $stockAfter]);

        StockLog::create([
            'raw_material_id' => $rawMaterialId,
            'user_id'         => $userId ?? auth()->id(),
            'type'            => $type,
            'quantity'        => $quantity,
            'stock_before'    => $stockBefore,
            'stock_after'     => $stockAfter,
            'reason'          => $reason,
            'notes'           => $notes,
        ]);

        return $material;
    }
}

File: lib/models/raw_material.dart
class RawMaterial {
  final int id;
  final String name;
  final double stock;
  final String unit;
  final double minStock;
  final String? image;
  final bool isActive;

  RawMaterial({
    required this.id,
    required this.name,
    required this.stock,
    required this.unit,
    required this.minStock,
    this.image,
    this.isActive = true,
  });

  factory RawMaterial.fromJson(Map<String, dynamic> json) {
    return RawMaterial(
      id: json['id'],
      name: json['name'],
      stock: double.tryParse(json['stock'].toString()) ?? 0.0,
      unit: json['unit'] ?? 'unit',
      minStock: double.tryParse(json['min_stock'].toString()) ?? 0.0,
      image: json['image'],
      isActive: json['is_active'] == 1 || json['is_active'] == true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'stock': stock,
      'unit': unit,
      'min_stock': minStock,
      'image': image,
      'is_active': isActive,
    };
  }
}
